File: cwiczenia 5 ocaml/zadania.ml




type 'a llist = LNil | LCons of 'a * (unit -> 'a llist);;

let rec lfrom k = LCons (k, function () -> lfrom (k+1));;

let rec ltake = function
| (0, _) -> []
| (_, LNil) -> []
| (n, LCons(x,xf)) -> x::ltake(n-1, xf());;


(* ZADANIE 1 *)
let lrepeat k lxs =
  let rec inLRepeat = function
    | (_, LNil) -> LNil
    | (0, LCons(x, xs)) -> inLRepeat (k, xs())
    | (n, (LCons(x, xs) as original)) -> LCons(x, function () -> inLRepeat (n-1, original))
  in inLRepeat(k, lxs);;


ltake (11,  lrepeat 2 (lfrom 0));;


(* ZADANIE 2 *)

let lfib =
  let rec inFib = function
      (pp, p) -> LCons(pp+p, fun () -> inFib (p, pp+p))
  in LCons(0, fun () -> LCons(1, fun () -> inFib (0, 1)));;

ltake (11,  lfib );;


(* ZADANIE 3 *)

type 'a lBT = LEmpty | LNode of 'a * (unit ->'a lBT) * (unit -> 'a lBT);;

let rec ltree n = LNode (n, (fun () -> ltree (2*n)), (fun () -> ltree (2*n + 1)));;

let breadthBT lTree = 
	let rec helper list=
		match list with
		| [] -> LNil
		| LEmpty::lTail -> helper lTail
		| LNode(x, lLeft, lRight)::lTail -> LCons(x, fun() -> helper(lTail@[lLeft();lRight()]) )
	in helper [lTree];;

ltake (20, breadthBT (ltree 1));;

File: cwiczenia 5 scala/src/zadania.scala


class zadania {
  
  
  
  // ZADANIE 1
  
def lrepeat (m:Int, ss:Stream[Int]):Stream[Int] = {
  def inRepeat (n:Int, s:Stream[Int]):Stream[Int]=
    (n,s) match {
    case (0, x#::xs) => inRepeat (m, xs)
    case (_, Stream.Empty) => Stream.Empty
    case (n, stream@x#::xs) => x#::inRepeat(n-1, stream)
  }//koniec match
  inRepeat (m, ss)
}//koniec metody
  
  val a = Stream(1,2,3)
  
  (lrepeat(3, Stream.from(4))).take(11).toList
  
  
  
  // ZADANIE 2
  
  def lfib = {
    def inFib (pp:Int, p:Int):Stream[Int] = (pp+p)#::inFib(p, pp+p)
    0#::1#::inFib(0,1)
  }
  
  lfib.take(10).toList
  
  //ZADANIE 3
  
  
 sealed trait lBT[+A]
 case object LEmpty extends lBT[Nothing]
 case class LNode[+A](elem:A, left:()=>lBT[A], right:()=>lBT[A]) extends lBT[A]

  
  def ltree (n:Int):lBT[Int] = LNode(n, () => ltree(2*n), () => ltree(2*n+1))
  
  def breadthBT (lbt:lBT[Int]):Stream[Int] = {
    def helper (list:List[lBT[Int]]):Stream[Int] = 
      list match {
      case List() => Stream.Empty
      case LEmpty::t => helper (t)
      case LNode(x, lf, rf)::t => x#::helper(t:::List(lf(), rf()))
    }//koniec match
    helper (List(lbt))
  }//koniec searcha
  
  
  
  breadthBT(ltree(1)).take(11).toList
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
}